src/bijectors/corr.jl

```diff
@@ -0,0 +1,135 @@
+# See stan doc for parametrization method:
+# https://mc-stan.org/docs/2_23/reference-manual/correlation-matrix-transform-section.html
+# (7/30/2020) their "manageable expression" is wrong...
+
+function upper1(AT, A)
+    AU = zero(AT)
+    for i=1:size(A,1), j=(i+1):size(A,2)
+        AU[i,j] = A[i,j]
+    end
+    AU
+end
+
+struct CorrBijector <: Bijector{2} end
+
+(b::CorrBijector)(X::AbstractMatrix{<:Real}) = link_lkj(X)
+(b::CorrBijector)(X::AbstractArray{<:AbstractMatrix{<:Real}}) = map(b, X)
+
+(ib::Inverse{<:CorrBijector})(Y::AbstractMatrix{<:Real}) = inv_link_lkj(Y)
+(ib::Inverse{<:CorrBijector})(Y::AbstractArray{<:AbstractMatrix{<:Real}}) = map(ib, Y)
+
+
+logabsdetjac(::Inverse{CorrBijector}, y::AbstractMatrix{<:Real}) = log_abs_det_jac_lkj(y)
+function logabsdetjac(b::CorrBijector, X::AbstractMatrix{<:Real})
+    
+    if !LinearAlgebra.isposdef(X)
+        println("!isposdef(X)")
+        return NaN # prevent Cholesky decomposition to break inference
```

Comment:
This will create a type instability. In general, we shouldn't check that the matrix is positive definite. IMO users should ensure positive definiteness if required.

Reply:
The Stan implementation for that occasionally broken transition is to reject this move and continue sampling. As I run following Stan code, written by @joshualeond
```stan
data {
  int<lower=1> N; // number of observations
  int<lower=1> J; // dimension of observations
  vector[J] y[N]; // observations
  vector[J] Zero; // a vector of Zeros (fixed means of observations)
}
parameters {
  corr_matrix[J] Omega; 
  vector<lower=0>[J] sigma; 
}
transformed parameters {
  cov_matrix[J] Sigma; 
  Sigma <- quad_form_diag(Omega, sigma); 
}
model {
  y ~ multi_normal(Zero,Sigma); // sampling distribution of the observations
  sigma ~ cauchy(0, 5); // prior on the standard deviations
  Omega ~ lkj_corr(1); // LKJ prior on the correlation matrix 
}
``` 

Few "error" info can be seen, like:
```
Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:

Informational Message: The current Metropolis proposal is about to be rejected because of the following issue:
Exception: correlation_model_namespace::log_prob: Sigma is not symmetric. Sigma[1,2] = inf, but Sigma[2,1] = inf (in '/home/yiyuezhuo/notebooks5/tmp/correlation.stan', line 12, column 2 to column 22)Exception: correlation_model_namespace::log_prob: Sigma is not symmetric. Sigma[1,2] = -inf, but Sigma[2,1] = -inf (in '/home/yiyuezhuo/notebooks5/tmp/correlation.stan', line 12, column 2 to column 22)

Exception: correlation_model_namespace::log_prob: Sigma is not symmetric. Sigma[1,2] = -inf, but Sigma[2,1] = -inf (in '/home/yiyuezhuo/notebooks5/tmp/correlation.stan', line 12, column 2 to column 22)
If this warning occurs sporadically, such as for highly constrained variable types like covariance matrices, then the sampler is fine,
If this warning occurs sporadically, such as for highly constrained variable types like covariance matrices, then the sampler is fine,
Exception: correlation_model_namespace::log_prob: Sigma is not symmetric. Sigma[1,2] = inf, but Sigma[2,1] = inf (in '/home/yiyuezhuo/notebooks5/tmp/correlation.stan', line 12, column 2 to column 22)
If this warning occurs sporadically, such as for highly constrained variable types like covariance matrices, then the sampler is fine,but if this warning occurs often then your model may be either severely ill-conditioned or misspecified.
but if this warning occurs often then your model may be either severely ill-conditioned or misspecified.

If this warning occurs sporadically, such as for highly constrained variable types like covariance matrices, then the sampler is fine,
but if this warning occurs often then your model may be either severely ill-conditioned or misspecified.
```

But the inference still runs and gets reasonable a result (I guess). But for Turing's [current implementation](https://github.com/TuringLang/Turing.jl/issues/702), such "error" will break inference entirely, so I call `isposdef` before `cholesky` to prevent entire failure.

Reply:
I think it's better to not perform these checks if possible since (depending on the employed matrix type) they perform a Cholesky decomposition to check if the matrix is positive definite.

IMO one could also thinl about a keword argument, similar to the one to `cholesky`, that allows to mute these errors. Additionally, since these errors are caused by numerical issues (?), a user could use algebraic transformations of `X` or e.g. PositiveFactorizarions.jl to avoid them. Also Bijectors is not written solely for Turing, so one should not only focus on its integration in Turing when designing this bijector.

Reply:
I see. If not considering Turing or HMC sampling integration, it is obvious that it should be removed. But I found that will lead to a very ugly hack to simulate Stan's reject behavior, which requires some strange value given by proposal at least will not halt the whole procedure. If so, some error should be mapped to NaN/Inf (ret) value instead of a vanilla fatal error.

Reply:
Good catch, and I agree. This is something that ought to be either handled by the user (by calling `isposdef` and making the logpdf `-Inf` to reject the sample) or inside Turing somehow, but not in Bijectors.

Reply:
Ok, I removed related code.